### PR TITLE
Support skip archive for Glue in Iceberg

### LINKED ISSUE #13413: Skip Glue archive in Iceberg table commits
like https://github.com/apache/iceberg/blob/3d00780f5f508ac7a0f475fd0e1ddfd736000c5b/aws/src/main/java/org/apache/iceberg/aws/glue/GlueTableOperations.java#L285
in https://github.com/trinodb/trino/blob/82a2cb6e5a160b392f539343400ca75b64d39ebf/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperations.java#L124

cc @danielcweeks @erichwang

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

`TestIcebergGlueCatalogSkipArchive.testSkipArchive` fails until updating IAM policy.
Fixes #13413

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Iceberg
* Add support for skipping archive when committing a table in Glue when
  the `iceberg.glue.skip-archive` configuration property is set to true. ({issue}`13413`)
```

### CODE REVIEW

**Comment by ebyhr** [comment]
@electrum Can you add `glue:GetTableVersions` action in IAM for `presto-ci` account? Sample screenshot:
Screen Shot 2022-09-28 at 11 51 34

**Comment by erichwang** [comment]
For what it's worth, I think we should just get rid of glue archiving altogether. Either that, or at some point we should make it default skip archiving. Otherwise, every Trino Iceberg cluster will blow up after so many thousands of transactions in Glue.

**Comment by ebyhr** [comment]
To support removing old archives, we need two more API calls (getTableVersions & batchDeleteTableVersion{Async}) & two more IAM permissions (GetTableVersions & BatchDeleteTableVersion). I lean toward enabling the flag by default in the future.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogConfig.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * limitations under the License.
+ */
+package io.trino.plugin.iceberg.catalog.glue;
+
+import io.airlift.configuration.Config;
+import io.airlift.configuration.ConfigDescription;
+
+public class IcebergGlueCatalogConfig
+{
+    private boolean skipArchive;
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogConfig.java:21]: Why `false` by default?

**Comment by findinpath** [comment]
Do you plan any follow-ups on hive/delta as well?

**Review by findinpath — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperationsProvider.java [hunk 2]
```diff
@@ -42,13 +42,14 @@ public GlueIcebergTableOperationsProvider(
...
             GlueHiveMetastoreConfig glueConfig,
+            IcebergGlueCatalogConfig icebergGlueConfig,
             AWSCredentialsProvider credentialsProvider)
     {
         this.fileSystemFactory = requireNonNull(fileSystemFactory, "fileSystemFactory is null");
         this.stats = requireNonNull(stats, "stats is null");
         requireNonNull(glueConfig, "glueConfig is null");
         requireNonNull(credentialsProvider, "credentialsProvider is null");
-        this.glueClient = createAsyncGlueClient(glueConfig, credentialsProvider, Optional.empty(), stats.newRequestMetricsCollector());
+        this.glueClient = createAsyncGlueClient(glueConfig, credentialsProvider, Optional.of(new SkipArchiveRequestHandler(icebergGlueConfig.isSkipArchive())), stats.newRequestMetricsCollector());
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperationsProvider.java:52]: Looks like we want to unify this with plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalogFactory.java by providing some Glue client provider interface, especially that `createAsyncGlueClient` has this parameter Optional

Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/GlueIcebergTableOperationsProvider.java]: Added a preparatory commit to unify them. Let me know if it's different from your intention.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogConfig.java (hunk 1 above)
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogConfig.java:21]: For backward compatibility. I can enable by default if there's no objection.

Comment (findepi) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogConfig.java:21]: it's unlikely that most people care about the Glue table versions for Iceberg tables, but it's also not unreasonable that someone uses that.

i think we should have it disabled (false) for now

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/SkipArchiveRequestHandler.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        this.skipArchive = skipArchive;
+    }
+
+    @Override
+    public AmazonWebServiceRequest beforeExecution(AmazonWebServiceRequest request)
+    {
+        if (request instanceof UpdateTableRequest updateTableRequest) {
+            return updateTableRequest.withSkipArchive(skipArchive);
+        }
+        return request;
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/SkipArchiveRequestHandler.java:36]: Glue Catalog uses only a couple of different requests.
Can we enumerate all these, to ensure this class gets updated if any new request variant is added?
for example, UpdateTableRequestV2 or something?

**Review by findepi — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogConfig.java (hunk 1 above)
Comment (electrum) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/IcebergGlueCatalogConfig.java:21]: Agreed, disabled by default makes sense.

**Comment by electrum** [comment]
@ebyhr sorry for the delay. Done now.

**Comment by colebow** [comment]
@ebyhr does this need to be included in docs as a new property, or is this just propagation of a property that already exists elsewhere?

**Comment by findinpath** [comment]
```
@Config("iceberg.glue.skip-archive")
@ConfigDescription("Skip archiving an old table version when creating a new version in a commit")
```

This one needs to be documented. (follow-up PR)

**Pull request merged**